soundness issue: uninitialized object variable of non-nullable type

I expect this code to give some kind of error:
```
class C { 
  method m() {
  }
}

method Main() {
  var c: C;
  c.m();
}
```

As you can see, the variable `c` will be `null` when `c.m()` is called.

However, it verifies with no issue, and when run, gives a `NullReferenceException`:

```
$ dafny foo.dfy 
Dafny 2.3.0.10506

Dafny program verifier finished with 1 verified, 0 errors
Compiled assembly into foo.exe

$ mono foo.exe

Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at _module.__default.Main () [0x00002] in <5ad62780ed7046e8ac2d47879dff4d86>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
  at _module.__default.Main () [0x00002] in <5ad62780ed7046e8ac2d47879dff4d86>:0 
```

--- a/Source/Dafny/Translator.cs
+++ b/Source/Dafny/Translator.cs
@@ -11770,8 +11770,11 @@ namespace Microsoft.Dafny {
       // Translate receiver argument, if any
       Expression receiver = bReceiver == null ? dafnyReceiver : new BoogieWrapper(bReceiver, dafnyReceiver.Type);
       if (!method.IsStatic && !(method is Constructor)) {
-        if (bReceiver == null && !(dafnyReceiver is ThisExpr)) {
-          CheckNonNull(dafnyReceiver.tok, dafnyReceiver, builder, etran, null);
+        if (bReceiver == null) {
+          TrStmt_CheckWellformed(dafnyReceiver, builder, locals, etran, true);
+          if (!(dafnyReceiver is ThisExpr)) {
+            CheckNonNull(dafnyReceiver.tok, dafnyReceiver, builder, etran, null);
+          }
         }
         ins.Add(etran.TrExpr(receiver));
       }

--- a/Test/git-issues/git-issue-448.dfy
+++ b/Test/git-issues/git-issue-448.dfy
@@ -1,0 +1,36 @@
+// RUN: %dafny /compile:3 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+class C {
+  var u: int
+  method M() {
+  }
+  function method F(): int {
+    5
+  }
+}
+
+codatatype Stream = More(x: int, tail: Stream)
+
+method Main() {
+  var c: C;
+  var s: Stream;
+  print s.x, "\n";  // error: "s" hasn't been initialized yet
+  if * {
+    // Regression test:
+    c.M();  // error: "c" hasn't been initialized yet
+  } else if * {
+    var f := c.F();  // error: "c" hasn't been initialized yet
+  } else if * {
+    P(c);  // error: "c" hasn't been initialized yet
+  } else if * {
+    var u := c.u;  // error: "c" hasn't been initialized yet
+  } else {
+    // Regression test:
+    (if 3 / 0 < 7 then c else c).M();  // error: division by zero
+  }
+}
+
+method P(w: C) {
+  w.M();
+}

--- a/Test/git-issues/git-issue-448.dfy.expect
+++ b/Test/git-issues/git-issue-448.dfy.expect
@@ -1,0 +1,21 @@
+git-issue-448.dfy(18,8): Error: variable 's', which is subject to definite-assignment rules, might be used before it has been assigned
+Execution trace:
+    (0,0): anon0
+git-issue-448.dfy(21,4): Error: variable 'c', which is subject to definite-assignment rules, might be used before it has been assigned
+Execution trace:
+    (0,0): anon0
+    (0,0): anon12_Then
+git-issue-448.dfy(23,13): Error: variable 'c', which is subject to definite-assignment rules, might be used before it has been assigned
+Execution trace:
+    (0,0): anon0
+    (0,0): anon13_Then
+git-issue-448.dfy(25,6): Error: variable 'c', which is subject to definite-assignment rules, might be used before it has been assigned
+Execution trace:
+    (0,0): anon0
+    (0,0): anon14_Then
+git-issue-448.dfy(27,13): Error: variable 'c', which is subject to definite-assignment rules, might be used before it has been assigned
+Execution trace:
+    (0,0): anon0
+    (0,0): anon15_Then
+
+Dafny program verifier finished with 1 verified, 5 errors

